src/coreclr/src/jit/lsra.cpp

```diff
@@ -2994,291 +2989,584 @@ regNumber LinearScan::tryAllocateFreeReg(Interval* currentInterval, RefPosition*
+            regMaskTP bestFitCandidateBit = genFindLowestBit(bestFitCandidates);
+            bestFitCandidates &= ~bestFitCandidateBit;
+            regNumber bestFitCandidateRegNum = genRegNumFromMask(bestFitCandidateBit);
+
+            // Find the next RefPosition of the register.
+            LsraLocation nextIntervalLocation = getNextIntervalRef(bestFitCandidateRegNum, regType);
+            LsraLocation nextPhysRefLocation  = getNextFixedRef(bestFitCandidateRegNum, regType);
+            nextPhysRefLocation               = Min(nextPhysRefLocation, nextIntervalLocation);
+            // If the nextPhysRefLocation is a fixedRef for the rangeEndRefPosition, increment it so that
+            // we don't think it isn't covering the live range.
+            // This doesn't handle the case where earlier RefPositions for this Interval are also
+            // FixedRefs of this regNum, but at least those are only interesting in the case where those
+            // are "local last uses" of the Interval - otherwise the liveRange would interfere with the reg.
+            // TODO: This duplicates code in an earlier loop, and is basically here to duplicate previous
+            // behavior; see if we can avoid this.
+            if (nextPhysRefLocation == rangeEndLocation && rangeEndRefPosition->isFixedRefOfReg(bestFitCandidateRegNum))
+            {
+                INDEBUG(dumpLsraAllocationEvent(LSRA_EVENT_INCREMENT_RANGE_END, currentInterval));
+                nextPhysRefLocation++;
+            }
+
+            if (nextPhysRefLocation == bestFitLocation)
+            {
+                bestFitSet |= bestFitCandidateBit;
+            }
+            else
+            {
+                bool isBetter = false;
+                if (nextPhysRefLocation > lastLocation)
+                {
+                    // This covers the full range; favor it if the other doesn't, or if it's a closer match.
+                    if ((bestFitLocation <= lastLocation) || (nextPhysRefLocation < bestFitLocation))
+                    {
+                        isBetter = true;
+                    }
+                }
+                else
+                {
+                    // This doesn't cover the full range; favor it if the other doesn't either, but this ends later.
+                    if ((bestFitLocation <= lastLocation) && (nextPhysRefLocation > bestFitLocation))
+                    {
+                        isBetter = true;
+                    }
+                }
+                if (isBetter)
+                {
+                    bestFitSet      = bestFitCandidateBit;
+                    bestFitLocation = nextPhysRefLocation;
+                }
+            }
+        }
+        assert(bestFitSet != RBM_NONE);
+        found = selector.applySelection(BEST_FIT, bestFitSet);
+    }
+
+    // Apply the IS_PREV_REG heuristic. Only applies if we have freeCandidates.
+    // Oddly, the previous heuristics only considered this if it covered the range.
```

Comment:
This just says that, before this refactoring, we would only prefer the previous register assignment if it covered the range, which seems odd to me because it seems like it would still be useful if all other conditions are the same. That is, if two registers have the same score, you'd probably want to prefer the previous register, even if neither of them cover the range. That *might* avoid a resolution move.